File: js/src/normalize.ts
export const ALIASES: Readonly<Record<string, string>> = {
  "gpt-5.1-latest": "gpt-5.1"
};

export const PROVIDER_PREFIXES: ReadonlySet<string> = new Set([
  "openai",
  "anthropic",
  "google",
  "xai",
  "meta",
  "mistral"
]);

function splitOnce(value: string, separator: string): [string, string] | null {
  const index = value.indexOf(separator);
  if (index <= 0 || index === value.length - separator.length) {
    return null;
  }
  return [value.slice(0, index), value.slice(index + separator.length)];
}

export function normalizeModelName(model: string): string {
  let normalized = model.trim().toLowerCase();
  if (normalized.length === 0) {
    throw new Error("model must not be empty");
  }

  const colonParts = splitOnce(normalized, ":");
  if (colonParts !== null && PROVIDER_PREFIXES.has(colonParts[0])) {
    normalized = colonParts[1];
  }

  const slashParts = splitOnce(normalized, "/");
  if (slashParts !== null && PROVIDER_PREFIXES.has(slashParts[0])) {
    normalized = slashParts[1];
  }

  return ALIASES[normalized] ?? normalized;
}

export function resolveModelKey(model: string, availableKeys: Iterable<string>): string | null {
  const keyMap = new Map<string, string>();
  for (const key of availableKeys) {
    keyMap.set(key.toLowerCase(), key);
    const normalizedKey = normalizeModelName(key);
    if (!keyMap.has(normalizedKey)) {
      keyMap.set(normalizedKey, key);
    }
  }

  const candidates: string[] = [];
  const raw = model.trim().toLowerCase();
  if (raw.length > 0) {
    candidates.push(raw);
  }
  candidates.push(normalizeModelName(model));

  for (const candidate of candidates) {
    const direct = keyMap.get(candidate);
    if (direct !== undefined) {
      return direct;
    }
    const alias = ALIASES[candidate];
    if (alias !== undefined) {
      const aliased = keyMap.get(alias);
      if (aliased !== undefined) {
        return aliased;
      }
    }
  }

  return null;
}

File: llmcalc/normalize.py
"""Model name normalization and lookup helpers."""

from __future__ import annotations

from collections.abc import Iterable

# Keep aliases small and explicit; extend as model IDs evolve.
ALIASES: dict[str, str] = {
    "gpt-5.1-latest": "gpt-5.1",
}

PROVIDER_PREFIXES: set[str] = {
    "openai",
    "anthropic",
    "google",
    "xai",
    "meta",
    "mistral",
}


def normalize_model_name(model: str) -> str:
    """Normalize an input model identifier to a canonical lookup key."""
    normalized = model.strip().lower()
    if not normalized:
        raise ValueError("model must not be empty")

    if ":" in normalized:
        provider, suffix = normalized.split(":", 1)
        if provider and suffix:
            normalized = suffix

    if "/" in normalized:
        provider, suffix = normalized.split("/", 1)
        if provider in PROVIDER_PREFIXES and suffix:
            normalized = suffix

    return ALIASES.get(normalized, normalized)


def resolve_model_key(model: str, available_keys: Iterable[str]) -> str | None:
    """Resolve a requested model id against available pricing table keys."""
    key_map: dict[str, str] = {}
    for key in available_keys:
        key_map[key.lower()] = key
        key_map.setdefault(normalize_model_name(key), key)

    candidates: list[str] = []
    raw = model.strip().lower()
    if raw:
        candidates.append(raw)
    candidates.append(normalize_model_name(model))

    for candidate in candidates:
        direct = key_map.get(candidate)
        if direct is not None:
            return direct
        alias = ALIASES.get(candidate)
        if alias is not None and alias in key_map:
            return key_map[alias]

    return None
